src/ot1d.jl

```diff
@@ -0,0 +1,283 @@
+"""
+function otCost1d(c,u::Vector,u_weights::Vector,v::Vector,v_weights::Vector)
+
+  @assert length(u) == length(u_weights)
+  @assert length(v) == length(v_weights)
+  @assert sum(u_weights) ≈ 1.0  atol=1e-7
+  @assert sum(v_weights) ≈ 1.0  atol=1e-7
+
+  cost = 0
+  perm_u = sortperm(u)
+  perm_v = sortperm(v)
+  u_sorted = u[perm_u]
+  v_sorted = v[perm_v]
+  u_w_sorted = u_weights[perm_u]
+  v_w_sorted = v_weights[perm_v]
+  
+  wi = u_w_sorted[1]
+  wj = v_w_sorted[1]
+  i,j = 1,1
+  while true
+    if (wi < wj || j == length(v_w_sorted))
+      cost += c(u_sorted[i],v_sorted[j])*wi
+      i+=1
+      if i == length(u_sorted)+1
+          break
+      end
+      wj -= wi
+      wi  = u_w_sorted[i]
+    else
+      cost += c(u_sorted[i],v_sorted[j])*wj
+      j+=1
+      if j == length(v_sorted)+1
+          break
+      end
+      wi -= wj
+      wj  = v_w_sorted[j]
+    end
+  end
+    
+  return cost
+end
+
+"""
+otPlan1d(c,u::Vector,u_weights::Vector,v::Vector,v_weights::Vector)
+otPlan1d(c,μ::Distributions.DiscreteNonParametric,ν::Distributions.DiscreteNonParametric)
+
+Calculates the Optimal Transport Plan between μ to ν, where
+they are 1-Dimensional distributions and the cost
+function is of the form ``c(x,y) = h(|x-y|)`` such that
+``h`` is a convex function.
+
+For the first case we have parameters:\n
+c: The cost function, which should be of the form ``c(x,y) = h(abs(x-y))``
+where ``h`` is a convex function. \n
+u        : Vector of point mass location    (e.g. `u = collect(1:10)`)\n
+u_weights: Vector of weights for each point (e.g. `u_weights = ones(10)*0.1`)\n
+v        : Vector of point mass location    (e.g. `v = collect(10:20)`)\n
+v_weights: Vector of weights for each point (e.g. `v_weights = ones(10)*0.1`)\n
+
+
+For the second case we have parameters:\n
+c: The cost function, which should be of the form ``c(x,y) = h(abs(x-y))``
+where ``h`` is a convex function. \n
+μ: Finite Discrete Distribution (e.g. `μ = DiscreteNonParametric(u, u_n)`)\n
+ν: Finite Discrete Distribution (e.g. `ν = DiscreteNonParametric(v, v_n)`)\n
+
+
+Returns the Optimal Transport Plan γ as a matrix.
+"""
+function otPlan1d(c,u::Vector,u_weights::Vector,v::Vector,v_weights::Vector)
+  γ = zeros(length(u),length(v))
+  perm_u = sortperm(u)
+  perm_v = sortperm(v)
+  u_sorted = u[perm_u]
+  v_sorted = v[perm_v]
+  u_w_sorted = u_weights[perm_u]
+  v_w_sorted = v_weights[perm_v]
+  
+  wi = u_w_sorted[1]
+  wj = v_w_sorted[1]
+  i,j = 1,1
+  while true
+      if (wi < wj || j == length(v_w_sorted))
+          γ[perm_u[i],perm_v[j]] = wi
+          i+=1
+          if i == length(u_sorted)+1
+              break
+          end
+          wj -= wi
+          wi  = u_w_sorted[i]
+      else
+          γ[perm_u[i],perm_v[j]] = wj
+          j+=1
+          if j == length(v_sorted)+1
+              break
+          end
+          wi -= wj
+          wj  = v_w_sorted[j]
+      end
+  end
+  
+  return γ
+end
```

Comment:
Same comments as above. Ideally, one would also not copy all the code from above.

Reply:
We could follow POT.py, and actually implement one function that calculates the plan and the cost behind the scenes, while
the `otcost1d` returns the first argument and `otplan1d` the second.